crux/src/Crux/Log.hs

```diff
@@ -25,23 +38,154 @@ module Crux.Log (
 
 import           Control.Exception ( SomeException, bracket_,  )
 import           Control.Lens
-import           Data.Text ( Text, lines )
+import           Control.Monad ( (<=<) )
+import           Data.Aeson ( Value(String), ToJSON, toJSON )
+import qualified Data.Text as T
 import           Data.Text.IO as TIO
+import           Data.Version ( Version, showVersion )
+import           Data.Word ( Word64 )
+import           GHC.Generics ( Generic )
 import qualified Lumberjack as LJ
+import           Prettyprinter
+import           Prettyprinter.Render.Text ( renderStrict )
 import           System.Console.ANSI
+import System.Console.Terminal.Size ( Window(Window), size )
 import           System.IO
 
 import           Lang.Crucible.Backend ( AssumptionReason )
 import qualified Lang.Crucible.Simulator.SimError as CSE
 
 import           Crux.Types
+import           Crux.Version ( version )
 
 ----------------------------------------------------------------------
 -- Various functions used by the main code to generate logging output
 
+newtype LogDoc = LogDoc (Doc ())
+
+instance ToJSON LogDoc where
+  toJSON (LogDoc doc) = String (renderStrict (layoutCompact doc))
+
+
+data CruxLogMessage ext
+  = AttemptingProvingVCs
+  | Checking [FilePath]
+  | ExtensionLogMessage ext
+  | Help LogDoc
+  | FoundCounterExample
+  | Goals [(T.Text, T.Text)]
+  | SimulationComplete
+  | SimulationTimedOut
+  | StartedGoal Integer
+  | TotalPathsExplored Word64
+  | Version T.Text Version
+  | WarnDisablingBranchCoverageRequiresPathSatisfiability
+  | WarnDisablingProfilingIncompatibleWithPathSplitting
+  | WarnPathsUnexplored Int
+  | WarnSkippingUnsatCoresBecauseMCSatEnabled
+  | WarnUnsupportedTimeoutFor String -- ^ name of the backend
+  deriving ( Generic, ToJSON )
+
+
+type SayWhatHandler ext =
+  ( ?extToSayWhat :: ext -> IO SayWhat
+  )
+
+
+cruxLogTag :: T.Text
+cruxLogTag = "Crux"
+
+
+cruxNoisily :: T.Text -> SayWhat
+cruxNoisily = SayWhat Noisily cruxLogTag
+
+cruxOK :: T.Text -> SayWhat
+cruxOK = SayWhat OK cruxLogTag
+
+cruxSimply :: T.Text -> SayWhat
+cruxSimply = SayWhat Simply cruxLogTag
+
+cruxWarn :: T.Text -> SayWhat
+cruxWarn = SayWhat Warn cruxLogTag
+
+
+cruxMessageToSayWhat :: SayWhatHandler ext => CruxLogMessage ext -> IO SayWhat
+
+cruxMessageToSayWhat AttemptingProvingVCs =
+  pure (cruxSimply "Attempting to prove verification conditions.")
```

Comment:
Very minor note: I prefer to only use `pure` in `Applicative`, but here it's `IO` which is a `Monad` so I would have used `return`.

Reply:
Ah, I tend to use `return` inside actually-effectful `do`-blocks, and `pure` when there really was nothing effectful going on.

Reply:
OK.  Like I said, it was just a very minor observation, so I'm fine with your reasoning.